File: implementations/cpp/external_sorting.cpp
/**
 * @file external_sorting.cpp
 * @brief Reference implementation of Two-Phase Multi-Way External Merge Sort.
 *
 * Implements Phase 1 (Run generation of size M) and Phase 2 (K-way merge via Min-Heap),
 * verifying correct sorted ordering and permutation invariance.
 * Verified with C++17 (-std=c++17 -O3 -Wall -Wextra -Werror).
 */

#include <iostream>
#include <vector>
#include <queue>
#include <cstdint>
#include <cassert>
#include <algorithm>
#include <random>

namespace dsa {

struct HeapEntry {
    int value;
    size_t run_idx;
    size_t element_idx;

    bool operator>(const HeapEntry& other) const {
        return value > other.value;
    }
};

/**
 * @brief Simulates Multi-Way External Merge Sort.
 * @param input The unsorted dataset (size N).
 * @param M RAM capacity (elements that can be sorted in memory at once).
 * @return The completely sorted dataset.
 */
inline std::vector<int> external_merge_sort(const std::vector<int>& input, size_t M) {
    assert(M > 0);
    size_t n = input.size();
    if (n <= M) {
        std::vector<int> sorted = input;
        std::sort(sorted.begin(), sorted.end());
        return sorted;
    }

    // --- Phase 1: Run Formation ---
    std::vector<std::vector<int>> runs;
    for (size_t i = 0; i < n; i += M) {
        size_t end_idx = std::min(i + M, n);
        std::vector<int> chunk(input.begin() + i, input.begin() + end_idx);
        std::sort(chunk.begin(), chunk.end());
        runs.push_back(std::move(chunk));
    }

    // --- Phase 2: K-Way Merge ---
    size_t k = runs.size();
    std::priority_queue<HeapEntry, std::vector<HeapEntry>, std::greater<HeapEntry>> min_heap;

    // Initialize heap with the first element of each run
    for (size_t r = 0; r < k; ++r) {
        if (!runs[r].empty()) {
            min_heap.push({runs[r][0], r, 0});
        }
    }

    std::vector<int> output;
    output.reserve(n);

    while (!min_heap.empty()) {
        HeapEntry top = min_heap.top();
        min_heap.pop();

        output.push_back(top.value);

        // Advance in the corresponding run
        size_t next_elem_idx = top.element_idx + 1;
        if (next_elem_idx < runs[top.run_idx].size()) {
            min_heap.push({runs[top.run_idx][next_elem_idx], top.run_idx, next_elem_idx});
        }
    }

    return output;
}

} // namespace dsa

int main() {
    std::cout << "Running Multi-Way External Merge Sort verification..." << std::endl;

    const size_t N = 10000;
    const size_t M = 250; // RAM can only hold 250 elements at a time -> 40 runs

    std::mt19937 rng(42);
    std::uniform_int_distribution<int> dist(-50000, 50000);

    std::vector<int> dataset(N);
    for (auto& x : dataset) x = dist(rng);

    // Run External Sort
    std::vector<int> sorted_result = dsa::external_merge_sort(dataset, M);

    // 1. Size invariance
    assert(sorted_result.size() == N);

    // 2. Ordering check
    for (size_t i = 1; i < N; ++i) {
        assert(sorted_result[i - 1] <= sorted_result[i]);
    }

    // 3. Oracle comparison against std::sort
    std::vector<int> oracle_sorted = dataset;
    std::sort(oracle_sorted.begin(), oracle_sorted.end());
    assert(sorted_result == oracle_sorted);

    std::cout << "[PASS] Successfully sorted " << N << " items with simulated RAM capacity M = " << M << "." << std::endl;
    std::cout << "[PASS] K-way merge result strictly matches std::sort oracle." << std::endl;
    std::cout << "All Multi-Way External Merge Sort assertions passed successfully!" << std::endl;
    return 0;
}

File: implementations/python/external_sorting.py
"""
Multi-Way External Merge Sort Simulation in Python.

Implements Phase 1 run generation and Phase 2 heapq-based multi-way merge
with unit tests comparing against standard sort.
"""

import heapq
import random
import unittest
from typing import List


def external_merge_sort(data: List[int], memory_capacity: int) -> List[int]:
    n = len(data)
    if n <= memory_capacity:
        return sorted(data)

    # Phase 1: Run Formation
    runs: List[List[int]] = []
    for i in range(0, n, memory_capacity):
        chunk = sorted(data[i : i + memory_capacity])
        runs.append(chunk)

    # Phase 2: K-way merge
    # Heap stores: (value, run_index, element_index)
    min_heap = []
    for r in range(len(runs)):
        if runs[r]:
            heapq.heappush(min_heap, (runs[r][0], r, 0))

    output: List[int] = []
    while min_heap:
        val, r_idx, elem_idx = heapq.heappop(min_heap)
        output.append(val)

        next_elem_idx = elem_idx + 1
        if next_elem_idx < len(runs[r_idx]):
            heapq.heappush(min_heap, (runs[r_idx][next_elem_idx], r_idx, next_elem_idx))

    return output


class TestExternalSorting(unittest.TestCase):
    def test_external_merge_sort_equivalence(self):
        rng = random.Random(42)
        dataset = [rng.randint(-10000, 10000) for _ in range(2500)]
        m = 100  # 25 runs

        sorted_res = external_merge_sort(dataset, m)
        self.assertEqual(sorted_res, sorted(dataset))


if __name__ == "__main__":
    unittest.main()
